File: sample/src/main/java/com/facebook/samples/litho/animations/AnimatedBadgeSpec.java
/*
 * This file provided by Facebook is for non-commercial testing and evaluation
 * purposes only.  Facebook reserves all rights not expressly granted.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
 * FACEBOOK BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN
 * ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN
 * CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

package com.facebook.samples.litho.animations;

import android.graphics.Color;
import android.graphics.Typeface;
import android.graphics.drawable.Drawable;
import android.graphics.drawable.ShapeDrawable;
import android.graphics.drawable.shapes.RoundRectShape;
import android.util.TypedValue;
import com.facebook.litho.ClickEvent;
import com.facebook.litho.Column;
import com.facebook.litho.Component;
import com.facebook.litho.ComponentContext;
import com.facebook.litho.Row;
import com.facebook.litho.StateValue;
import com.facebook.litho.Transition;
import com.facebook.litho.animation.AnimatedProperties;
import com.facebook.litho.annotations.LayoutSpec;
import com.facebook.litho.annotations.OnCreateLayout;
import com.facebook.litho.annotations.OnCreateTransition;
import com.facebook.litho.annotations.OnEvent;
import com.facebook.litho.annotations.OnUpdateState;
import com.facebook.litho.annotations.State;
import com.facebook.litho.widget.Text;
import com.facebook.yoga.YogaAlign;
import com.facebook.yoga.YogaEdge;
import java.util.Arrays;

@LayoutSpec
public class AnimatedBadgeSpec {
  private static final int ANIMATION_DURATION = 1000;
  private static final Transition.TransitionAnimator ANIMATOR =
      Transition.timing(ANIMATION_DURATION);

  private static final String TRANSITION_KEY_OUTER_CONTAINER = "outer_container";
  private static final String TRANSITION_KEY_CONTAINER = "container";
  private static final String TRANSITION_KEY_TEXT = "text";
  private static final String TRANSITION_KEY_BADGE = "badge";

  @OnCreateLayout
  static Component onCreateLayout(ComponentContext c, @State boolean showText) {
    return Row.create(c)
        .child(Column.create(c).paddingDip(YogaEdge.ALL, 8).child(buildComment(c, showText)))
        .build();
  }

  static Component buildComment(ComponentContext c, boolean showText) {
    return Column.create(c)
        .transitionKey(TRANSITION_KEY_OUTER_CONTAINER)
        .paddingDip(YogaEdge.ALL, 8)
        .child(
            Row.create(c)
                .alignItems(YogaAlign.CENTER)
                .child(
                    Text.create(c)
                        .textSizeSp(16)
                        .textStyle(Typeface.BOLD)
                        .text("Cristobal Castilla"))
                .child(
                    Row.create(c)
                        .transitionKey(TRANSITION_KEY_CONTAINER)
                        .marginDip(YogaEdge.LEFT, 8)
                        .paddingDip(YogaEdge.ALL, 3)
                        .alignItems(YogaAlign.CENTER)
                        .child(
                            Column.create(c)
                                .heightDip(18)
                                .widthDip(18)
                                .background(buildRoundedRect(c, 0xFFFFB74B, 9)))
                        .child(
                            !showText
                                ? null
                                : Text.create(c)
                                    .transitionKey(TRANSITION_KEY_TEXT)
                                    .marginDip(YogaEdge.LEFT, 8)
                                    .clipToBounds(true)
                                    .textSizeDip(12)
                                    .text("Top Follower"))
                        .child(
                            Text.create(c)
                                .transitionKey(TRANSITION_KEY_BADGE)
                                .marginDip(YogaEdge.LEFT, 8)
                                .marginDip(YogaEdge.RIGHT, 4)
                                .textSizeDip(12)
                                .textColor(Color.BLUE)
                                .text("+1"))
                        .clickHandler(AnimatedBadge.onClick(c))
                        .background(buildRoundedRect(c, Color.WHITE, 12))))
        .child(Text.create(c).textSizeSp(18).text("So awesome!"))
        .background(buildRoundedRect(c, 0xFFDDDDDD, 20))
        .build();
  }

  @OnEvent(ClickEvent.class)
  static void onClick(ComponentContext c) {
    AnimatedBadge.toggleVisibility(c);
  }

  @OnUpdateState
  static void toggleVisibility(StateValue<Boolean> showText) {
    showText.set(!showText.get());
  }

  @OnCreateTransition
  static Transition onCreateTransition(ComponentContext c) {
    return Transition.parallel(
        Transition.create(TRANSITION_KEY_CONTAINER, TRANSITION_KEY_OUTER_CONTAINER)
            .animate(AnimatedProperties.WIDTH)
            .animator(ANIMATOR),
        Transition.create(TRANSITION_KEY_BADGE).animate(AnimatedProperties.X).animator(ANIMATOR),
        Transition.create(TRANSITION_KEY_TEXT)
            .animate(AnimatedProperties.WIDTH)
            .appearFrom(0f)
            .disappearTo(0f)
            .animator(ANIMATOR),
        Transition.create(TRANSITION_KEY_TEXT)
            .animate(AnimatedProperties.ALPHA)
            .appearFrom(0f)
            .disappearTo(0f)
            .animator(ANIMATOR));
  }

  private static Drawable buildRoundedRect(ComponentContext c, int color, int cornerRadiusDp) {
    final float cornerRadiusPx =
        TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP, cornerRadiusDp, c.getResources().getDisplayMetrics());
    final float[] radii = new float[8];
    Arrays.fill(radii, cornerRadiusPx);
    final RoundRectShape roundedRectShape = new RoundRectShape(radii, null, radii);
    final ShapeDrawable drawable = new ShapeDrawable(roundedRectShape);
    drawable.getPaint().setColor(color);
    return drawable;
  }
}

File: sample-kotlin/src/main/kotlin/com/facebook/samples/litho/kotlin/animations/animatedbadge/AnimatedBadgeSpec.kt
/*
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.facebook.samples.litho.kotlin.animations.animatedbadge

import android.graphics.Color
import android.graphics.Typeface
import com.facebook.litho.ClickEvent
import com.facebook.litho.Column
import com.facebook.litho.Component
import com.facebook.litho.ComponentContext
import com.facebook.litho.Row
import com.facebook.litho.StateValue
import com.facebook.litho.Transition
import com.facebook.litho.Transition.BaseTransitionUnitsBuilder
import com.facebook.litho.animation.AnimatedProperties
import com.facebook.litho.annotations.LayoutSpec
import com.facebook.litho.annotations.OnCreateLayout
import com.facebook.litho.annotations.OnCreateTransition
import com.facebook.litho.annotations.OnEvent
import com.facebook.litho.annotations.OnUpdateState
import com.facebook.litho.annotations.State
import com.facebook.litho.widget.Text
import com.facebook.samples.litho.kotlin.drawable.RoundedRect
import com.facebook.yoga.YogaAlign
import com.facebook.yoga.YogaEdge
import com.facebook.yoga.YogaPositionType

@LayoutSpec
object AnimatedBadgeSpec {

  private const val ANIMATION_DURATION = 300
  private val ANIMATOR = Transition.timing(ANIMATION_DURATION)

  private const val TRANSITION_KEY_TEXT = "text"

  @OnCreateLayout
  fun onCreateLayout(c: ComponentContext, @State state: Int): Component {
    val expanded1 = state == 1 || state == 2
    val expanded2 = state == 2 || state == 3
    return Column.create(c)
        .paddingDip(YogaEdge.ALL, 8f)
        .child(Row.create(c).marginDip(YogaEdge.TOP, 8f).child(buildComment1(c, expanded1)))
        .child(Row.create(c).marginDip(YogaEdge.TOP, 16f).child(buildComment2(c, expanded2)))
        .clickHandler(AnimatedBadge.onClick(c))
        .build()
  }

  private fun buildComment1(c: ComponentContext, expanded: Boolean): Component =
      Column.create(c)
          .paddingDip(YogaEdge.ALL, 8f)
          .child(
              Row.create(c)
                  .alignItems(YogaAlign.CENTER)
                  .child(
                      Text.create(c)
                          .textSizeSp(16f)
                          .textStyle(Typeface.BOLD)
                          .text("Cristobal Castilla"))
                  .child(
                      Row.create(c)
                          .marginDip(YogaEdge.LEFT, 8f)
                          .paddingDip(YogaEdge.ALL, 3f)
                          .alignItems(YogaAlign.CENTER)
                          .child(
                              Column.create(c)
                                  .heightDip(18f)
                                  .widthDip(18f)
                                  .background(RoundedRect.build(c, -0x48b5, 9)))
                          .child(
                              if (!expanded) {
                                null
                              } else {
                                Text.create(c)
                                    // still need transition keys for appear/disappear animations
                                    .transitionKey(TRANSITION_KEY_TEXT)
                                    .transitionKeyType(Transition.TransitionKeyType.GLOBAL)
                                    // need this to prevent the global key of "+1" Text from
                                    // changing
                                    .key("text")
                                    .marginDip(YogaEdge.LEFT, 8f)
                                    .clipToBounds(true)
                                    .textSizeDip(12f)
                                    .text("Top Follower")
                              })
                          .child(
                              Text.create(c)
                                  .marginDip(YogaEdge.LEFT, 8f)
                                  .marginDip(YogaEdge.RIGHT, 4f)
                                  .textSizeDip(12f)
                                  .textColor(Color.BLUE)
                                  .text("+1"))
                          .background(RoundedRect.build(c, Color.WHITE, 12))))
          .child(Text.create(c).textSizeSp(18f).text("So awesome!"))
          .background(RoundedRect.build(c, -0x222223, 20))
          .build()

  private fun buildComment2(c: ComponentContext, expanded: Boolean): Component =
      Column.create(c)
          .paddingDip(YogaEdge.ALL, 8f)
          .child(
              Row.create(c)
                  .alignItems(YogaAlign.CENTER)
                  .child(
                      Text.create(c)
                          .textSizeSp(16f)
                          .textStyle(Typeface.BOLD)
                          .text("Cristobal Castilla"))
                  .child(
                      Row.create(c)
                          .widthDip((if (expanded) 48 else 24).toFloat())
                          .marginDip(YogaEdge.LEFT, 8f)
                          .paddingDip(YogaEdge.ALL, 3f)
                          .alignItems(YogaAlign.CENTER)
                          .child(
                              Column.create(c)
                                  .positionType(YogaPositionType.ABSOLUTE)
                                  .positionDip(YogaEdge.LEFT, (if (expanded) 27 else 3).toFloat())
                                  .heightDip(18f)
                                  .widthDip(18f)
                                  .background(RoundedRect.build(c, 0xFFB2CFE5.toInt(), 9)))
                          .child(
                              Column.create(c)
                                  .positionType(YogaPositionType.ABSOLUTE)
                                  .positionDip(YogaEdge.LEFT, (if (expanded) 15 else 3).toFloat())
                                  .heightDip(18f)
                                  .widthDip(18f)
                                  .background(RoundedRect.build(c, 0xFF4B8C61.toInt(), 9)))
                          .child(
                              Column.create(c)
                                  .heightDip(18f)
                                  .widthDip(18f)
                                  .background(RoundedRect.build(c, 0xFFFFB74B.toInt(), 9)))
                          .background(RoundedRect.build(c, Color.WHITE, 12))))
          .child(Text.create(c).textSizeSp(18f).text("So awesome!"))
          .background(RoundedRect.build(c, 0xFFDDDDDD.toInt(), 20))
          .build()

  @OnEvent(ClickEvent::class)
  fun onClick(c: ComponentContext) {
    AnimatedBadge.updateState(c)
  }

  @OnUpdateState
  fun updateState(state: StateValue<Int>) {
    state.set((state.get()!! + 1) % 4)
  }

  @OnCreateTransition
  fun onCreateTransition(c: ComponentContext): Transition =
      Transition.parallel<BaseTransitionUnitsBuilder>(
          Transition.allLayout().animator(ANIMATOR),
          Transition.create(Transition.TransitionKeyType.GLOBAL, TRANSITION_KEY_TEXT)
              .animate(AnimatedProperties.WIDTH)
              .appearFrom(0f)
              .disappearTo(0f)
              .animator(ANIMATOR),
          Transition.create(Transition.TransitionKeyType.GLOBAL, TRANSITION_KEY_TEXT)
              .animate(AnimatedProperties.ALPHA)
              .appearFrom(0f)
              .disappearTo(0f)
              .animator(ANIMATOR))
}
